protocol/xmrtaker/swap_state.go

```diff
@@ -154,13 +123,100 @@ func newSwapState(
 		exchangeRate,
 		ethAsset,
 		stage,
-		walletScanHeight,
+		moneroStartNumber,
 		statusCh,
 	)
 	if err = b.SwapManager().AddSwap(info); err != nil {
 		return nil, err
 	}
 
+	defer func() {
+		statusCh <- stage
+	}()
+	return newSwapState(
+		b,
+		transferBack,
+		info,
+		ethHeader.Number,
+		moneroStartNumber,
+	)
+}
+
+func newSwapStateFromOngoing(
+	b backend.Backend,
+	info *pswap.Info,
+	transferBack bool,
+	ethSwapInfo *db.EthereumSwapInfo,
+	sk *mcrypto.PrivateKeyPair,
+) (*swapState, error) {
+	if info.Status != types.ETHLocked && info.Status != types.ContractReady {
+		return nil, errInvalidStageForRecovery
+	}
+
+	makerSk, makerVk, err := b.RecoveryDB().GetXMRMakerSwapKeys(info.ID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get xmrmaker swap keys from db: %w", err)
+	}
+
+	s, err := newSwapState(
+		b,
+		transferBack,
+		info,
+		ethSwapInfo.StartNumber,
+		info.MoneroStartHeight,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if b.ContractAddr() != ethSwapInfo.ContractAddress {
+		return nil, errContractAddrMismatch(ethSwapInfo.ContractAddress.String())
+	}
+
+	s.setTimeouts(ethSwapInfo.Swap.Timeout0, ethSwapInfo.Swap.Timeout1)
+	s.privkeys = sk
+	s.pubkeys = sk.PublicKeyPair()
+	s.contractSwapID = ethSwapInfo.SwapID
+	s.contractSwap = ethSwapInfo.Swap
+	s.xmrmakerPublicSpendKey = makerSk
+	s.xmrmakerPrivateViewKey = makerVk
+	return s, nil
+}
+
+func newSwapState(
+	b backend.Backend,
+	transferBack bool,
+	info *pswap.Info,
+	ethStartNumber *big.Int,
+	moneroStartNumber uint64,
+) (*swapState, error) {
+	_, err := b.XMRDepositAddress(nil)
+	if transferBack && err != nil {
+		return nil, errMustProvideWalletAddress
+	}
```

Comment:
In theory this shouldn't be possible.  We automatically create a Monero wallet if it doesn't exist.  Transfer back is the default. We talked about removing alternate deposit addresses as complexity that might do more damage than good (since money can be lost if the address is wrong), but there was concern that a UI might let the user set an alternate deposit address that would only work as a taker and not as a maker. I would have nuked the `XMRDepositAddress` function, but I didn't want to force a rushed decision. Now that some time has passed, what do you think: Is the use case of an alternate XMR deposit address desirable enough to justify the extra code and non-orthogonal maker/taker Monero handling?

Reply:
I think for the UI, users will need to specify the address for the monero to go to - is there currently an option at the moment for them to do that via CLI? I don't think there is, other than if they make the wallet beforehand and point to that.

Reply:
No, is it a flag that needs to be added to `swapcli take ...`?